variables.tf

```diff
@@ -1,36 +1,300 @@
+  default     = "Writer"
 }
 
-variable "plan" {
+# variable "plan" {
+#   description = "The plan to use when Object Storage instances are created. [Learn more](https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-provision)."
+#   type        = string
+#   default     = "standard"
+#   validation {
+#     condition     = contains(["standard", "cos-one-rate-plan"], var.plan)
+#     error_message = "The specified plan is not a valid selection!"
+#   }
+# }
+
+variable "cos_location" {
+  description = "The location for the Object Storage instance."
   type        = string
-  description = "The name of the plan type supported by service."
+  default     = "global"
+}
+
+##############################################################################
+# KMS instance variables
+##############################################################################
+
+variable "key_protect_name" {
+  type        = string
+  description = "The name of the Key Protect instance to create."
+  default     = "key-protect"
+}
+
+variable "kp_plan" {
+  type        = string
+  description = "Plan for the Key Protect instance. Valid plans are 'tiered-pricing' and 'cross-region-resiliency', for more information on these plans see [Key Protect pricing plan](https://cloud.ibm.com/docs/key-protect?topic=key-protect-pricing-plan)."
+  default     = "tiered-pricing"
+
+  validation {
+    condition     = contains(["tiered-pricing", "cross-region-resiliency"], var.kp_plan)
+    error_message = "`plan` must be one of: 'tiered-pricing', 'cross-region-resiliency'."
+  }
+
+  validation {
+    condition     = var.kp_plan == "tiered-pricing" ? true : (var.kp_plan == "cross-region-resiliency" && contains(["us-south", "eu-de", "jp-tok"], var.region))
+    error_message = "'cross-region-resiliency' is only available for the following regions: 'us-south', 'eu-de', 'jp-tok'."
+  }
+}
+
+variable "allowed_network" {
+  type        = string
+  description = "Allowed networks for the Key Protect instance. Possible values: 'private-only', 'public-and-private'."
+  default     = "public-and-private"
+
+  validation {
+    condition     = can(regex("public-and-private|private-only", var.allowed_network))
+    error_message = "Valid values for allowed_network are 'public-and-private' or 'private-only'."
+  }
+}
+
+variable "standard_key" {
+  type        = bool
+  description = "Whether to create a standard key (true) or an imported key (false)."
+  default     = false
+}
+
+variable "endpoint_type" {
+  type        = string
+  description = "The type of endpoint for the KMS key. Options: 'public', 'private'."
+  default     = "public"
+}
+
+##############################################################################
+# COS bucket variables
+##############################################################################
+
+variable "bucket_name" {
+  type        = string
+  description = "The name for the new Object Storage bucket."
```

Comment:
this should mention that if user is passing a prefix, it will be added infront of this